django/db/models/query.py

```diff
@@ -1619,8 +1620,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
 
         if single:
             val = vals[0] if vals else None
-            to_attr = to_attr if as_attr else cache_name
-            setattr(obj, to_attr, val)
+
+            if as_attr:
+                # A to_attr has been given for the prefetch
+                setattr(obj, to_attr, val)
+            elif is_descriptor:
+                # cache_name points to a field name in obj
+                # this field is likely a descriptor for a related object.
+                setattr(obj, cache_name, val)
+            else:
+                # No to_attr has been given for this prefetch operation
+                # Using "cache_name" combined with "is_descriptor" means
```

Comment:
I think this comment needs updating to reflect the reversal of sense from the old `is_field` to the new `is_descriptor`